docs/arch/adr-003-versioning-branching-strategy.md

```diff
@@ -0,0 +1,66 @@
+
+We need a branching strategy to support the goals of the versioning strategy. We do not want to have to stop development on `main` as we await a release, nor do we want every change in `main` published in every release. Using branches, we can decide which changes go into which release.
+
+## Requirements
+The following are requirements that must be met for releasing dbt:
+
+1. We must have major, minor, and patch releases.
+1. We must have release types that have the goal of garnering early feedback from the Community.
+1. We must have release types that aim to identify bugs and test for stability.
+1. We must have release types that are stable and of production quality.
+1. We must have the ability to isolate changes from different versions based on the type of release.
+
+## Decisions
+
+### Version Types
+<details>
+<summary> Header Explanations: </summary>
+
+- Type: the type of the version (e.g. beta, final)
+- Released?: does this type of version need to be released?
+- Branches: the branches where these version types are present (e.g. beta version numbers should only exist on the `main` branch)
+- Release Versions: the releases where the version type is applicable (i.e. we will only have alphas for major and minor releases, we will not have alphas for patch releases)
+- Expectations: the stability of the code changes in the release
+</details>
+
+| Type | Released? | Branches | Release Versions | Expectations |
+| ---- | --------- | -------- | ---------------- | ------------ |
+| Alpha | No | `main` | Major / Minor | Experimental |
+| Beta | Yes | `main` | Major / Minor | Experimental |
+| RC | Yes | Release branch | Major / Minor / Patch | Pre-production |
+| Final | Yes | Release branch | Major / Minor / Patch | Production |
+
+### Branching Strategy
+Based off of the release version and expectations, the version type's branching strategy can be determined. 
+ * Alpha and Beta versions are experimental, and therefore exist only on the `main` branch, where all our changes reside
+ * Alpha versions are never released. They signify that development is underway, though the changes they contain have not been released in any form
+ * An RC version denotes a more stable state. A release branch will be created for RC releases to limit the changes a release contains
+ * Final versions are stable and tested. Only verified changes will go into these releases, which are inherited from the corresponding RC version
+ * A release branch will exist for each unique major or minor version, to be named accordingly: `<major>.<minor>.latest` (e.g. `1.0.latest`)
```

Comment:
I think we should consider creating a release branch for every release down to the patch level.  There have been at least few cases where that would have made life much easier.

Reply:
I have also liked working on projects that have a 1:1 relationship between releases and branches.

Reply:
@iknox-fa can you explain what we gain by doing it this way?  It seems to me it adds complexity of what branch to merge to.  It also adds complexity to the backport action and our changelog generation.

Reply:
@emmyoop the way I've done it, you cut a branch for the release then never touch it again. No more figuring out what commit a release was released from and rolling back your local git to investigate bugs.

Reply:
@iknox-fa or @nathaniel-may have you then just created patch branches off of the initial minor version branch? So for example: 
For the beta, I create the `1.1.0` branch off of `main`
Then for `1.1.1` patch, I create the `1.1.1` branch off of the `1.1.0` branch?
What about for RCs? Do they get their own branch as well?

Reply:
@leahwicz exactly like that. I'm used to all releases having a branch including RCs.

Reply:
@iknox-fa @nathaniel-may we tag the commit that is released. Are tags not a good way to map what we released to what version? Relying solely on a branch for having only the commits we released actually scare me since in theory you can add more commits to them after a release (not that it would be intentional) but a tag is a definitive end commit.

Reply:
tags do work for this mapping (I wrote more about it in our slack thread). But the same worry exists with our current strategy because we could edit the history of the branch. It might be a little harder to do by accident, but we already live with that possibility. In both situations we would have branch protections to limit the surface area of that risk.